src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,493 @@
+          let !br' = BB.BufferRange (op `plusPtr` inpRemaining) ope
+          k br'
+      | otherwise = do
+          SBS.copyToPtr sbs ip op outRemaining
+          let !ip' = ip + outRemaining
+          return $ BB.bufferFull 1 ope (go ip' ipe)
+      where
+        outRemaining = ope `minusPtr` op
+        inpRemaining = ipe - ip
+
+{-------------------------------------------------------------------------------
+  Accumulator for page contents
+-------------------------------------------------------------------------------}
+
+-- We might have preferred a representation for an incrementally constructed
+-- page that already places the raw bytes in a page-sized buffer. However, we do
+-- not know a priori what the offsets for most of the page components will be.
+-- These offsets depend on the number of keys\/values\/blob references, and the
+-- sizes of keys\/values, and we only find out these numbers as key-operation
+-- pairs are added incrementally. As such, we can only construct the raw page
+-- bytes once the current page has filled up.
+--
+-- TODO: an alternative representation that would put less load on the GC would
+-- be to use mutable vectors for each component. Where we currently only copy
+-- bytes once (from input to the output page), we would have to copy each byte
+-- one extra time to the intermediate mutable vectors. However, it would save
+-- having to allocate and de-allocate heap objects, like lists. We wouldn't know
+-- a priori exactly how large those vectors should be, but we know they are
+-- bounded, though special care should be taken for multi-page values.
+data PageAcc = PageAcc {
+    -- | (1) directory of components
+    pageSize           :: !PageSize
+    -- | (2) an array of 1-bit blob reference indicators
+  , pageBlobRefBitmap  :: !BitMap
+    -- | (3) an array of 2-bit operation types
+  , pageOperations     :: !CrumbMap
+    -- | (4) a pair of arrays of blob references
+  , pageBlobRefOffsets :: !(StricterList Word64)
+    -- | (4) a pair of arrays of blob references, ctd
+  , pageBlobRefSizes   :: !(StricterList Word32)
+    --   (5) key offsets will be computed when serialising the page
+    --   (6) value offsets will be computed when serialising the page
+    -- | (7) the concatenation of all keys
+  , pageKeys           :: !(StricterList RawKey)
+    -- | (8) the concatenation of all values
+  , pageValues         :: !(StricterList RawValue)
+  }
+
+paIsEmpty :: PageAcc -> Bool
+paIsEmpty p = psIsEmpty (pageSize p)
+
+paIsOverfull :: PageAcc -> Bool
+paIsOverfull p = psIsOverfull (pageSize p)
+
+paEmpty :: PageAcc
+paEmpty = PageAcc {
+      pageSize           = psEmpty
+    , pageBlobRefBitmap  = emptyBitMap
+    , pageOperations     = emptyCrumbMap
+    , pageBlobRefOffsets = SNil
+    , pageBlobRefSizes   = SNil
+    , pageKeys           = SNil
+    , pageValues         = SNil
+    }
+
+paAddElem ::
+     KOp
+  -> PageAcc
+  -> Maybe PageAcc
+paAddElem kop PageAcc{..}
```

Comment:
After looking at #55 again, this function is probably where we should ensure that the range finder bits of all entries of the page match. Or do you have a different approach in mind?

Reply:
You are quite right -- I'm going to add this